server/policy.go

```diff
@@ -13,7 +19,69 @@ import (
 // will describe in what way the state failed. See the Policy documentation for
 // more information about the specifics of different policies.
 func EvaluatePolicy(state *pb.MachineState, policy *pb.Policy) error {
-	return evaluatePlatformPolicy(state.GetPlatform(), policy.GetPlatform())
+	if err := evaluatePlatformPolicy(state.GetPlatform(), policy.GetPlatform()); err != nil {
+		return err
+	}
+	if state.GetTeeAttestation() == nil {
+		return nil
+	}
+	switch at := state.TeeAttestation.(type) {
+	case *pb.MachineState_SevSnpAttestation:
+		return evaluateSevSnpPolicy(at.SevSnpAttestation, policy.GetSevSnp())
+	case *pb.MachineState_TdxAttestation:
+		// Currently unchecked.
+		return nil
+	default:
+		return fmt.Errorf("no policy for TEE attestation type %T", state.TeeAttestation)
+	}
+}
+
+func rootOfTrust(certs [][]byte) (*x509.CertPool, error) {
+	pool := x509.NewCertPool()
+	for _, der := range certs {
+		cert, err := x509.ParseCertificate(der)
+		if err != nil {
+			return nil, err
+		}
+		pool.AddCert(cert)
+	}
+	return pool, nil
+}
+
+// the SEV-SNP attestation signature is already verified by this point.
+func evaluateSevSnpPolicy(state *spb.Attestation, policy *pb.SevSnpPolicy) error {
+	if policy == nil {
+		return nil
+	}
+	if state == nil {
+		return fmt.Errorf("attestation is nil")
+	}
+	// No UEFI policy. Done.
+	if policy.GetUefi() == nil {
+		return nil
+	}
+	// Extract which certs to trust as root for keys that sign uefi measurements
+	uefirot, err := rootOfTrust(policy.GetUefi().GetRootCerts())
+	if err != nil {
+		return err
+	}
+	kind := svalidate.CertEntryKind(svalidate.CertEntryAllowMissing)
+	if policy.GetUefi().GetRequire() {
+		kind = svalidate.CertEntryRequire
+	}
+	// Check a SEV-SNP attestation's extra certificate table for signed uefi measurements
+	// and apply the verification logic against the attestation measurement.
+	opts := &svalidate.Options{
+		GuestPolicy: defaultSevSnpGuestPolicy,
+		CertTableOptions: map[string]*svalidate.CertEntryOption{
+			gcesev.GCEFwCertGUID: {
+				Kind: kind,
+				Validate: gceverify.SNPValidateFunc(&gceverify.Options{
+					RootsOfTrust: uefirot,
+					Now:          time.Now(),
+					// TODO: No getter? Network fallback can be helpful.
```

Comment:
What specifically do we need to do here? Is it to introduce a Getter/an option for a Getter?

Reply:
How do you feel about `EvaluatePolicyOpt` with a new `*PolicyOptions` argument? The `EvaluatePolicy` can use default options like `Getter: trust.DefaultHTTPSGetter()` and `Now: time.Now()` (Now moved to options to make evaluateSevSnpPolicy more deterministic and testable).